### PR TITLE
Ongoing OIDC: Add support for Scope.PrimaryAuthMethodId

### PR DESCRIPTION
**Pull request by jimlambrt**
In order to incorporate the new Scope.PrimaryAuthMethodId,  this PR includes:

1. The schema changes needed to add `primary_auth_method_id` to the `iam_scope` table.
2.  Changes to `iam.Repository` needed to support `Scope.PrimaryAuthMethodId`
3.  Fix all the bits across Boundary that had `iam.WithAutoVivify(true)` hard coded and remove the option from Boundary.
4. Support for updating the PrimaryAuthMethod in the scope's service handler
5. A new `-primary-auth-method-id` option for the CLI: `scopes update` command.
6. A new README which attempts to capture at a very high level list of things to-do when you're adding an additional field to an existing API and corresponding CLI command... like `Scope.PrimaryAuthMethodId` for instance.  The README's name isn't great, so I'm open to suggestions.
7. Some clean-up for transitioning operational states for an OIDC auth method, which was broken by a recent merge.
8. Export of authtoken's `NewAuthTokenId()` which is in an open un-reviewed PR and included in this PR as a dup, but required to make bits compile and make tests complete.
 
@mgaffney, @talanknight: I'd appreciate a complete review by both of you, with some attention to the bits in the service, CLI and of course the repo and schema changes.

@jefferai: I didn't make any changes to the initial resources, since the global auth password method doesn't need to be set as the primary auth method for the system to work properly... and I didn't think it was a good idea.

@randallmorey: I've included you as a reviewer, since these changes have an impact on the FE. I think, we're already aligned on them, but I'm happy to discuss.

@s-christoff, @louisruch: I added you as optional reviewers. I welcome your feedback on any of this.  I'm curious if the you find the new readme helpful as well.

### CODE REVIEW

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto [hunk 1]
```diff
@@ -10,55 +10,65 @@ import "google/protobuf/struct.proto";
...
 
-	// Output only. The ID of the parent Scope, if any. This field will be empty if this is the "global" scope.
-	string parent_scope_id = 5 [json_name="parent_scope_id"];
+  // Output only. The ID of the parent Scope, if any. This field will be empty if this is the "global" scope.
+  string parent_scope_id = 5 [json_name = "parent_scope_id"];
+
+  // Output only. The ID of the primary auth method for this scope. A primary auth method
+  // is allowed to vivify users when new accounts are created and is the source for
+  // the users account info
+  string primary_auth_method_id = 6;
```
Comment (talanknight) [inline internal/proto/local/controller/api/resources/scopes/v1/scope.proto:31]: Please add `[json_name = "primary_auth_method_id"]`

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto [hunk 2]
```diff
@@ -10,55 +10,65 @@ import "google/protobuf/struct.proto";
...
-	// The mutation will fail if the version does not match the latest known good version.
-	uint32 version = 80;
+  // The type of the resource.
+  string type = 90;
 
-	// The type of the resource.
-	string type = 90;
+  // The ID of the primary auth method for this scope.  A primary auth method
+  // is allowed to vivify users when new accounts are created and is the source for the users account info
+  google.protobuf.StringValue primary_auth_method_id = 100
```
Comment (talanknight) [inline internal/proto/local/controller/api/resources/scopes/v1/scope.proto:66]: Please add [json_name = "primary_auth_method_id"]

File: internal/servers/controller/handlers/scopes/scope_service.go [hunk 3]
```diff
@@ -318,6 +318,9 @@ func (s Service) updateInRepo(ctx context.Context, parentScope *scopes.ScopeInfo
 	if name := item.GetName(); name != nil {
 		opts = append(opts, iam.WithName(name.GetValue()))
 	}
+	if primaryAuthMethodId := item.GetPrimaryAuthMethodId(); primaryAuthMethodId != nil {
```
Comment (talanknight) [inline internal/servers/controller/handlers/scopes/scope_service.go:321]: Consider adding a check in validateUpdateRequest along the lines of:

```
if item.GetPrimaryAuthMethodId() != nil  && !handlers.ValidId(item.GetPrimaryAuthMethodId().GetValue(), password.AuthMethodPrefix, oidc.AuthMethodPrefix) {
  badFields["primary_auth_method_id"] = "Improperly formatted identifier"
}
```

This type of ValidId check (with multiple possible prefixes) might require https://github.com/hashicorp/boundary/pull/977 to be merged.

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto (hunk 1 above)
Comment (talanknight) [reply internal/proto/local/controller/api/resources/scopes/v1/scope.proto:31]: ScopeInfo provides info regarding a resource's parent scope.   I'm having a hard time thinking of a situation where knowing a parent scope's primary auth method would be helpful, but maybe you or @randallmorey have already identified a situation?

**Review by talanknight — COMMENTED** [review]
Looks good!

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto (hunk 1 above)
Comment (jimlambrt) [reply internal/proto/local/controller/api/resources/scopes/v1/scope.proto:31]: Hum... I agree that it's not useful, so I'm removing it.

Comment (jimlambrt) [reply internal/proto/local/controller/api/resources/scopes/v1/scope.proto:31]: ty!

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto (hunk 2 above)
Comment (jimlambrt) [reply internal/proto/local/controller/api/resources/scopes/v1/scope.proto:66]: done.

File: internal/servers/controller/handlers/scopes/scope_service.go (hunk 3 above)
Comment (jimlambrt) [reply internal/servers/controller/handlers/scopes/scope_service.go:321]: added todo for dependent PR merge.

File: internal/servers/controller/handlers/scopes/scope_service.go [hunk 4]
```diff
@@ -318,6 +318,14 @@ func (s Service) updateInRepo(ctx context.Context, parentScope *scopes.ScopeInfo
 	if name := item.GetName(); name != nil {
 		opts = append(opts, iam.WithName(name.GetValue()))
 	}
+	// TODO (jimlambrt 3/2021) uncomment this ValidId(...) check once https://github.com/hashicorp/boundary/pull/977
```
Comment (talanknight) [inline internal/servers/controller/handlers/scopes/scope_service.go:321]: This is ok to use now.

**Review by talanknight — APPROVED** [review]

File: internal/cmd/commands/scopescmd/funcs.go [hunk 5]
```diff
@@ -9,6 +9,12 @@ import (
 	"github.com/hashicorp/boundary/internal/cmd/base"
 )
 
+const (
```
Comment (jefferai) [inline internal/cmd/commands/scopescmd/funcs.go:12]: Good thought. I should have done this when refactoring but just didn't think about it. We could update that over time though.

File: internal/cmd/commands/scopescmd/funcs.go [hunk 6]
```diff
@@ -17,30 +23,38 @@ func init() {
...
-				Name:   "skip-default-role-creation",
+				Name:   flagSkipDefaultRoleCreationName,
 				Target: &c.flagSkipDefaultRoleCreation,
 				Usage:  "If set, a role granting the anonymous user access to log into auth methods and a few other actions within the newly-created scope will not automatically be created",
 			})
+		case flagPrimaryAuthMethodIdName:
+			f.StringVar(&base.StringVar{
+				Name:   flagPrimaryAuthMethodIdName,
+				Target: &c.flagPrimaryAuthMethodId,
+				Usage:  "If set, the primary auth method id for the scope.  A primary auth method is allowed to autovivify users and is used as a source for account full name and email for a scope's users",
```
Comment (jefferai) [inline internal/cmd/commands/scopescmd/funcs.go:56]: I think most people won't know what autovivify means. I would change it to "create users on first login" or so.

**Comment by jefferai** [comment]
> @jefferai: I didn't make any changes to the initial resources, since the global auth password method doesn't need to be set as the primary auth method for the system to work properly... and I didn't think it was a good idea.

I think you mean this because we create everything automatically, right? But won't it then not create users if you create new password accounts using that auth method and they log in? In which case we probably should indeed be marking that initial password method as the primary for the scope.

File: internal/iam/repository_user.go [hunk 7]
```diff
@@ -306,6 +311,26 @@ func (r *Repository) LookupUserWithLogin(ctx context.Context, accountId string,
...
+	if acct == nil {
+		return false, errors.New(errors.InvalidParameter, op, "missing account")
+	}
+	acctScope := allocScope()
+	acctScope.PublicId = acct.ScopeId
+	err := r.reader.LookupByPublicId(context.Background(), &acctScope)
+	if err != nil {
+		return false, errors.Wrap(err, op, errors.WithMsg(fmt.Sprintf("unable to lookup account's scope %s", acct.ScopeId)))
+	}
+	switch {
```
Comment (jefferai) [inline internal/iam/repository_user.go:326]: This switch is fine but you could also just simplify it to `return acct.AuthMethodId == acctScope.PrimaryAuthMethodId, nil` if you like.

File: internal/proto/local/controller/api/resources/scopes/v1/scope.proto [hunk 8]
```diff
@@ -10,55 +10,60 @@ import "google/protobuf/struct.proto";
...
-	string type = 90;
+  // The type of the resource.
+  string type = 90;
 
-	// Output only. The available actions on this resource for this user.
-	repeated string authorized_actions = 300 [json_name="authorized_actions"];
+  // The ID of the primary auth method for this scope.  A primary auth method
+  // is allowed to vivify users when new accounts are created and is the source for the users account info
+  google.protobuf.StringValue primary_auth_method_id = 100
+      [json_name = "primary_auth_method_id", (custom_options.v1.generate_sdk_option) = true, (custom_options.v1.mask_mapping) = { this: "primary_auth_method_id" that: "PrimaryAuthMethodId" }];
```
Comment (jefferai) [inline internal/proto/local/controller/api/resources/scopes/v1/scope.proto:62]: TIL this could be on a new line.

**Review by jefferai — APPROVED** [review]
Made a few comments (both in the code and one on the PR itself) but hitting approve cause it's really good otherwise!

File: internal/cmd/commands/scopescmd/funcs.go (hunk 6 above)
Comment (jimlambrt) [reply internal/cmd/commands/scopescmd/funcs.go:56]: ty!

File: internal/iam/repository_user.go (hunk 7 above)
Comment (jimlambrt) [reply internal/iam/repository_user.go:326]: ty!

File: internal/servers/controller/handlers/scopes/scope_service.go (hunk 4 above)
Comment (jimlambrt) [reply internal/servers/controller/handlers/scopes/scope_service.go:321]: ty!

**Comment by jimlambrt** [comment]
@jefferai - I moved forward with making the initial password auth method the primary auth method for the global scope, so users will be created on first time login.  I didn't add any logic for making the first auth method of a scope, the primary auth method for the new scope.  Happy add that in a future PR, if/when we decide to go that route.

@talanknight - please review the new changes to `scopes.(Service).updateInRepo(...)`.  It didn't appear that we were properly handling updates to the global scope.  It looked like they were being treated as on update to an Org scope.

File: internal/servers/controller/handlers/scopes/scope_service.go [hunk 9]
```diff
@@ -312,20 +314,39 @@ func (s Service) createInRepo(ctx context.Context, authResults auth.VerifyResult
...
+		opts = append(opts, iam.WithPrimaryAuthMethodId(scopePrimaryAuthMethodId))
 	}
 	version := item.GetVersion()
 
 	var iamScope *iam.Scope
 	var err error
-	switch parentScope.GetType() {
-	case scope.Global.String():
+	switch {
+	case scopeId == scope.Global.String():
```
Comment (talanknight) [inline internal/servers/controller/handlers/scopes/scope_service.go:338]: please add a test case which updates the name and/or description of the global scope.

**Review by talanknight — CHANGES_REQUESTED** [review]

File: internal/servers/controller/handlers/scopes/scope_service.go [hunk 10]
```diff
@@ -312,20 +314,39 @@ func (s Service) createInRepo(ctx context.Context, authResults auth.VerifyResult
...
+		scopeDesc = desc.GetValue()
+		opts = append(opts, iam.WithDescription(scopeDesc))
 	}
 	if name := item.GetName(); name != nil {
-		opts = append(opts, iam.WithName(name.GetValue()))
+		scopeName = name.GetValue()
+		opts = append(opts, iam.WithName(scopeName))
+	}
+	if primaryAuthMethodId := item.GetPrimaryAuthMethodId(); primaryAuthMethodId != nil {
+		if !handlers.ValidId(primaryAuthMethodId.GetValue(), password.AuthMethodPrefix, oidc.AuthMethodPrefix) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/controller/handlers/scopes/scope_service.go:327]: I misunderstood your previous change request.  I'll move it.  ty!

**Review by talanknight — APPROVED** [review]

**Pull request merged**